### PR TITLE
Introduce watermarks for indexing pressure backoff

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently we have a relatively basic decider about when to throttling
indexing. This commit adds two levels of watermarks with configurable
bulk size deciders. Additionally, adds additional settings to control
primary, coordinating, and replica rejection limits.

### CODE REVIEW

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/internalClusterTest/java/org/elasticsearch/action/bulk/IncrementalBulkIT.java [hunk 1]
```diff
@@ -180,6 +183,57 @@ public void testIncrementalBulkRequestMemoryBackOff() throws Exception {
...
+        }
+
+        ArrayList<DocWriteRequest<?>> requests = new ArrayList<>();
+        add512BRequests(requests, index);
+
+        CountDownLatch finishLatch = new CountDownLatch(1);
+        blockWritePool(threadPool, finishLatch);
+        IncrementalBulkService.Handler handler = incrementalBulkService.newBulkRequest();
+        refCounted.incRef();
+        handler.addItems(requests, refCounted::decRef, () -> nextPage.set(true));
```
[the remark this answers is not in the document]
Comment (Tim-Brooks) [reply server/src/internalClusterTest/java/org/elasticsearch/action/bulk/IncrementalBulkIT.java]: I made modifications to one of the tests to include this.

File: server/src/main/java/org/elasticsearch/index/IndexingPressure.java [hunk 2]
```diff
@@ -62,14 +105,22 @@ public class IndexingPressure {
...
     private final AtomicLong primaryDocumentRejections = new AtomicLong(0);
 
-    private final long primaryAndCoordinatingLimits;
-    private final long splitBulkThreshold;
+    private final long lowWatermark;
+    private final long lowWatermarkSize;
+    private final long highWatermark;
+    private final long highWatermarkSize;
+    private final long coordinatingLimits;
+    private final long primaryLimits;
```
[the remark this answers is not in the document]
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/index/IndexingPressure.java]: Just copy and pasted the old value which was plural for some reason. I made them non-plural.

**Comment by Tim-Brooks** [comment]
Backporting is waiting for: https://github.com/elastic/elasticsearch/pull/113940.

**Pull request merged**